### PR TITLE
Make Launching Apps With Existing Databases Easier

### LINKED ISSUE #1719: `fly launch` doesn't notice when it can't create a PG app due to the name being taken
fly v0.0.458 darwin/arm64 Commit: bf6a2deb BuildDate: 2023-02-15T18:33:43Z

To reproduce:

* Follow the Getting Started guide for Elixir at https://fly.io/docs/elixir/getting-started/ to create and launch a new app (say, `myapp`).
* Delete the app (pretend you messed up and just want to start over) but forget to delete its database app. Now there's a Fly Postgres called `myapp-db`.
* Launch a fresh app; call it `myapp` again. Say "Y" to creating a Postgres database. Choose a cluster size.
* Watch the launch and deploy look happy until the release command fails because DATABASE_URL is missing.

```
Set secrets on cphx2: SECRET_KEY_BASE
? Would you like to set up a Postgresql database now? Yes
? Select configuration: Development - Single node, 1x shared CPU, 256MB RAM, 1GB disk
Creating postgres cluster in organization personal
Creating app...
? Would you like to set up an Upstash Redis database now? No
Preparing system for Elixir builds
Installing application dependencies
Running Docker release generator
Wrote config file fly.toml
? Would you like to deploy now? Yes
...
==> Release command detected: /app/bin/migrate
...
	 Preparing to run: `/app/bin/migrate` as nobody
	 2023/02/15 22:23:59 listening on [fdaa:0:3b99:a7b:88db:e7e0:a4d0:2]:22 (DNS: [fdaa::3]:53)
	 ERROR! Config provider Config.Reader failed with:
	 ** (RuntimeError) environment variable DATABASE_URL is missing.
	 For example: ecto://USER:PASS@HOST/DATABASE
	     (elixir 1.14.0) src/elixir.erl:298: anonymous fn/4 in :elixir.eval_external_handler/1
	     (stdlib 4.0.1) erl_eval.erl:492: :erl_eval.expr/6
	     (stdlib 4.0.1) erl_eval.erl:136: :erl_eval.exprs/6
	     (elixir 1.14.0) src/elixir.erl:288: :elixir.eval_forms/3
	 {"init terminating in do_boot",{#{'__exception__'=>true,'__struct__'=>'Elixir.RuntimeError',message=><<101,110,118,105,114,111,110,109,101,110,116,32,118,97,114,105,97,98,108,101,32,68,65,84,65,66,65,83,69,95,85,82,76,32,105,115,32,109,105,115,115,105,110,103,46,10,70,111,114,32,101,120,97,109,112,108,101,58,32,101,99,116,111,58,47,47,85,83,69,82,58,80,65,83,83,64,72,79,83,84,47,68,65,84,65,66,65,83,69,10>>},[{elixir_eval,'__FILE__',1,[{file,"/app/releases/0.1.0/runtime.exs"},{line,26}]},{elixir,'-eval_external_handler/1-fun-2-',4,[{file,"src/elixir.erl"},{line,298},{error_info,#{module=>'Elixir.Exception'}}]},{erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,748}]},{erl_eval,expr,6,[{file,"erl_eval.erl"},{line,492}]},{erl_eval,exprs,6,[{file,"erl_eval.erl"},{line,136}]},{elixir,eval_forms,3,[{file,"src/elixir.erl"},{line,288}]},{'Elixir.Module.ParallelChecker',verify,1,[{file,"lib/module/parallel_checker.ex"},{line,100}]},{'Elixir.Code',validated_eval_string,3,[{file,"lib/code.ex"},{line,422}]}]}}
	 init terminating in do_boot ({,[{elixir_eval,__FILE__,1,[{_},{_}]},{elixir,-eval_external_handler/1-fun-2-,4,[{_},{_},{_}]},{erl_eval,do_apply,7,[{_},{_}]},{erl_eval,expr,6,[{_},{_}]},{erl_eval,exprs,6,[{_},{_}]},{elixir,eval_forms,3,[{_},{_}]},{Elixir.Module.ParallelChecker,verify,1,[{_},{_}]},{Elixir.Code,validated_eval_string,3,[{_},{_}]}]})
	 Crash dump is being written to: erl_crash.dump...done
	 Starting clean up.
Error release command failed, deployment aborted
```

Zoom in on some debug output:

```
? Would you like to set up a Postgresql database now? Yes
? Select configuration: Development - Single node, 1x shared CPU, 256MB RAM, 1GB disk
Creating postgres cluster in organization personal
Creating app...
DEBUG --> POST https://api.fly.io/graphql

{
  "query": "mutation($input: CreateAppInput!) { createApp(input: $input) { app { id name organization { slug } config { definition } regions { name code } } } }",
  "variables": {
    "input": {
      "organizationId": "bjkZRb7BXzk36H383wbX0z9bXRH6LL",
      "name": "cphx2-db",
      "preferredRegion": "yyz",
      "appRoleId": "postgres_cluster"
    }
  }
}

DEBUG {}
DEBUG <-- 200 https://api.fly.io/graphql (223.64ms)

{
  "data": {
    "createApp": null
  },
  "errors": [
    {
      "message": "Validation failed: Name has already been taken",
      "locations": [
        {
          "line": 1,
          "column": 37
        }
      ],
      "path": [
        "createApp"
      ],
      "extensions": {
        "code": "UNPROCESSABLE"
      }
    }
  ]
}
? Would you like to set up an Upstash Redis database now? (y/N)
```

### PR DESCRIPTION
**Pull request by billyb2**
Should fix #1719 

This PR does two things. Firstly, it gives users the option of destroying an app's database at the same time as the app itself, which should make the scenario that an app's database already exists a little harder to hit. It also warns the user when it fails to launch the PG or Redis database, so that users no launching an app with an existing database could pose problems for the launch.

Related discussion:
https://flyio.discourse.team/t/destroying-an-apps-database-in-fly-apps-destroy/1902

Edit:
After some discussion, the PR currently allows users to attach an existing database, if a database exists that matches "{app_name}-db". It also warns users when failing to launch a database (redis or PG) that this could cause issues during deployment.

### CODE REVIEW

File: internal/command/launch/launch.go [hunk 1]
```diff
@@ -349,13 +351,69 @@ func run(ctx context.Context) (err error) {
...
+				}
+
+			}
+
 			options["postgresql"] = true
+
+			if should_attach_db {
+				current_time := time.Now().Nanosecond()
+				db_user := fmt.Sprintf("%s-%d", db_app_name, current_time)
+				fmt.Println(db_user)
```
Comment (tvdfly) [inline internal/command/launch/launch.go:379]: How about removing this, or adding context?

File: internal/command/launch/launch.go [hunk 2]
```diff
@@ -349,13 +351,69 @@ func run(ctx context.Context) (err error) {
...
+				err = postgres.AttachCluster(ctx, postgres.AttachParams{
+					PgAppName: db_app_name,
+					AppName:   appConfig.AppName,
+					DbUser:    db_user,
+				})
+
+				if err != nil {
+					msg := `Failed attaching %s to the Postgres cluster %s: %w.\nTry attaching manually with 'fly postgres attach --app %s %s'\n`
```
Comment (tvdfly) [inline internal/command/launch/launch.go:387]: what do you think about offering to stop `fly launch` if postgres was requested and we failed to attach? for example, a yes/no prompt that asks?

I know for some frameworks they automatically make sqlite databases if they don't find database settings in the environment, so developers might like the opportunity to bail out if this step fails.

File: internal/command/launch/launch.go (hunk 1 above)
Comment (billyb2) [reply internal/command/launch/launch.go:379]: Definitely should add context. If we just try to use the typical username generation format of using the app's name ,when we'll run into the issue of not having a unique username. By using a sufficiently random number (in this case, the nano second that this command is being run at), we can pretty much guarantee we're creating a unique user. I'll add a comment about this.

File: internal/command/launch/launch.go (hunk 2 above)
Comment (billyb2) [reply internal/command/launch/launch.go:387]: That sounds like a good idea. Do you think we should do it for launching postgres as well?

**Review by tvdfly — APPROVED** [review]
this is sooo much nicer, thank you!

```
? Would you like to set up a Postgresql database now? Yes
? We found an existing Postgresql database with the name long-silence-7192-db. Would you like to attach it to your app? Yes
Checking for existing attachments
? Database "long_silence_7192" already exists. Continue with the attachment process? Yes
Registering attachment
Creating user

Postgres cluster long-silence-7192-db is now attached to long-silence-7192
The following secret was added to long-silence-7192:
  DATABASE_URL=postgres://long_silence_7192_db_777999000:...@top2.nearest.of.long-silence-7192-db.internal:5432/long_silence_7192?sslmode=disable
Postgres cluster long-silence-7192-db is now attached to long-silence-7192
```

**Pull request merged**